Topics/Tech_Stacks/Metaprogramming.md

```diff
@@ -0,0 +1,142 @@
+to_json_message and from_json_message methods in the Message class.
+
+The magic happens in the message class from which all other message types inherit.
+
+```python
+
+class Message(BaseModel):
+    #this is the method that generates the json serialization.
+    # for example class SignIn(username: str, password: str) would return
+    # {"SignIn": {"username": "username", "password": "password"}}
+    def to_json_message(self) -> str:
+        # we find the name of the class through the __class__ attribute.
+        name = self.__class__.__name__
+        # we convert the class to a dictionary using the model_dump method.
+        data = self.model_dump_json()
+        # then we add the class name as a key to the dictionary. and convert the dictionary to json.
+        return json.dumps({name: data})
+
+    # this is the method that generates the json deserialization and returns a 
+    # specific Message object that we can use
+    # for dynamic dispatch.
+    # for example {"SignIn": {"username": "username", "password": "password"}} 
+    # would return a SignIn object with the username and password fields set.
+    @staticmethod
+    def from_json_message(message: str) -> Message:
+        # we convert the json to a dictionary.
+        dictionary = json.loads(message)
+        # extract the class name and the class fields from the dictionary.
+        name, data = next(iter(dictionary.items()))
+        data = json.loads(data)
+        # we look up the class in the global scope and return a new instance of 
+        # the class with the data as the arguments.
+        cls = globals()[name]
+        return cls(**data)
+
+    # this is the method that handles the logic of the message. We need to implement
+    # this method in all our Message subclasses.
+    # for example class SignIn(username: str, password: str) would return a SignInResponse object.
+    # this is how we get dynamic dispatch to work. We can call the respond method on any Message object and it will return the correct response. as defined by the subclass.
+    def respond(self) -> Message:
+        raise NotImplementedError
+```
+
+Then assuming we have a server object that can send and receive JSON. 
+the server architecture is as follows.
+
+```python
+    while True:
+        json: str = server.get_json()
+        # we convert the json to a Message object using the from_json_message method.
+        message_as_cls: Message = Message.from_json_message(json)
+        # we use dynamic dispatch to call the respond method of the Message object.
+        response_as_cls: Message = message_as_cls.respond()
+        # we convert the response to json and send it back to the client.
+        response_json = response_cls.to_json_message()
+        server.send_json(response_json)
+```
+
+remember to import at the top of the file.
+
+```python
+from __future__ import annotations
+from datetime import datetime
+from typing import *
+import json
+
+from pydantic import BaseModel
```

Comment:
Explain what this import does perhaps? And why it's necessary in this conext?

Reply:
added a deatiled comment on every import